### ISSUE TITLE
Possible performance increase in `random_agent`

### ISSUE THREAD

**Issue by Datseris** [JuliaDynamics/Agents.jl#742]
What if this instead first filters the iterator returned by `allids` for valid agents, then chooses a random one from the result? We avoid allocating `collect(allids(model))` but end up iterating through the entire list. I imagine most `condition` would be simple checks (e.g. `agent.group == 1`), so the allocation cost might outweigh the iteration cost.

_Originally posted by @AayushSabharwal in https://github.com/JuliaDynamics/Agents.jl/pull/741#discussion_r1107564606_

**Comment by Datseris** [comment]
(with respect to the current implementation:
```julia
function random_agent(model, condition)
    ids = shuffle!(abmrng(model), collect(allids(model)))
    i, L = 1, length(ids)
    a = model[ids[1]]
    while !condition(a)
        i += 1
        i > L && return nothing
        a = model[ids[i]]
    end
    return a
end
```
)

**Comment by jacobusmmsmit** [comment]
The problem with this is that if you do a non-allocating lazy filter with `Iterators.filter`, you don't know how many elements the generator has a priori, and so you can't select an element from it using a random number with some implementation like
```julia
function random_agent(model, condition)
    ids = Iterators.filter(condition, allagents(model)
    idx = rand(abmrng(model)
    return ids[idx]
end
```

**Comment by Datseris** [comment]
Yeap, this doesn't work.

**Comment by Datseris** [comment]
To clarify. The original suggestion that made me open this issue simply doesn't work. The code can't really become better.

**Comment by Datseris** [comment]
Or, to be more precise: _maybe there is a way to make the quoted function faster, but this is definitely not the original suggestion._

**Comment by Datseris** [comment]
We can add a keyword that allows to first get a random agent and then check its condition, and then continue in a `while` loop until we get an agent that actually satisfies condition. If most of the agents of the model will satisfy the condition, then the version that calls `random_agent(model)` and then checks will probably be faster due to not allocating. That seems like an easy code to set up and test @jacobusmmsmit if you're up for it

**Comment by jacobusmmsmit** [comment]
I'll give it a go.

**Comment by Tortar** [comment]
Solved in #744
